src/ConnectionRequest.jl

```diff
@@ -23,20 +23,60 @@ abstract type ConnectionPoolLayer{Next <: Layer} <: Layer end
 export ConnectionPoolLayer
 
 function request(::Type{ConnectionPoolLayer{Next}}, url::URI, req, body;
+                 proxy=nothing,
                  socket_type::Type=TCPSocket, kw...) where Next
 
+    if proxy != nothing
+        target_url = url
+        url = URI(proxy)
+        if target_url.scheme == "http"
+            req.target = string(target_url)
+        end
+    end
+
     IOType = ConnectionPool.Transaction{sockettype(url, socket_type)}
     io = getconnection(IOType, url.host, url.port; kw...)
 
     try
+        if proxy != nothing && target_url.scheme == "https"
+            return tunnel_request(Next, io, target_url, req, body; kw...)
+        end
+
         return request(Next, io, req, body; kw...)
+
     catch e
         @debug 1 "❗️  ConnectionLayer $e. Closing: $io"
         close(io)
         rethrow(isioerror(e) ? IOError(e, "during request($url)") : e)
+    finally
+        if proxy != nothing && target_url.scheme == "https"
+            close(io)
+        end
     end
+
+end
+
+sockettype(url::URI, default) = url.scheme in ("wss", "https") ? SSLContext :
+                                                                 default
+
+function tunnel_request(Next, io, target_url, req, body; kw...)
+    r = connect_tunnel(io, target_url)
+    if r.status != 200
+        return r
+    end
+    io = ConnectionPool.sslupgrade(io, target_url.host; kw...)
+    return request(Next, io, req, body; kw...)
+end
+
+function connect_tunnel(io, target_url)
+    target = "$(URIs.hoststring(target_url.host)):$(target_url.port)"
+    @debug 1 "📡  CONNECT HTTPS tunnel to $target"
+    request = Request("CONNECT", target)
+    writeheaders(io, request)
+    startread(io)
+    readheaders(io, request.response)
+    return request.response
```

Comment:
To proxy HTTPS, we set the target to "host:port" and send a `CONNECT` request to establish a tunnel to the target. If that fails, just return the non-200 Response object. It will be converted into an exception by the ExceptionLayer, retried, etc as usual. If we get a 200 response, call `sslupgrade` to change the `Transaction{TCPSocket}` into a `Transaction{SSLContext}`.